src/Plugin/EntityReferenceSelection/OgSelection.php

```diff
@@ -77,37 +77,49 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
       $query->condition($bundle_key, $bundles, 'IN');
     }
 
-    $user_groups = $this->getUserGroups();
-    if (!$user_groups) {
+    $user = User::load(\Drupal::currentUser()->id());
+
+    if (!$user) {
+      // @todo: What to do with anonymous user?
       return $query;
     }
 
+    if ($user->hasPermission('administer group')) {
+      // User can see all the groups.
+      return $query;
+    }
+
+    // User is a non-site wide group admin.
     $identifier_key = $definition->getKey('id');
 
+    /** @var \Drupal\og\OgAccessInterface $og_access */
+    $og_access = \Drupal::service('og.access');
```

Comment:
How can you inject service to field constraints validator? I've looked for that and could not find anything.